beaker/session.py

```diff
@@ -366,14 +458,106 @@ def invalidate(self):
+                self.clear()
+                try:
+                    session_data = self.namespace2['session']
+
+                    if (session_data is not None and self.encrypt_key):
+                        session_data = self._decrypt_data(session_data, migration=True)
+                    elif not self.old_reads():
+                        # Memcached always returns a key, its None when its not
+                        # present
+                        session_data = {
+                            '_creation_time': now,
+                            '_accessed_time': now
+                        }
+                        self.is_new = True
+                        read_value = True
+
+                except (KeyError, TypeError):
+                    session_data = None
+                    if self.old_reads():
+                        # We still have another backend we could be reading from, so don't create new sessions here
+                        pass
+                    else:
+                        # Post migration we should 
+                        session_data = {
+                            '_creation_time': now,
+                            '_accessed_time': now
+                        }
+                        self.is_new = True
+                        read_value = True
+
+                if not self.old_reads() and (session_data is None or len(session_data) == 0):
+                    session_data = {
+                        '_creation_time': now,
+                        '_accessed_time': now
+                    }
+                    self.is_new = True
+                    read_value = True
+
+                # Only consider the case where we successfully read a session
+                if session_data is None or len(session_data) == 0:
+                    pass
+                elif self.timeout is not None and \
+                    now - session_data['_accessed_time'] > self.timeout:
+                    timed_out = True
+                    read_value = True
+                else:
+                    read_value = True
+                    # Properly set the last_accessed time, which is different
+                    # than the *currently* _accessed_time
+                    if self.is_new or '_accessed_time' not in session_data:
+                        self.last_accessed = None
+                    else:
+                        self.last_accessed = session_data['_accessed_time']
+
+                    # Update the current _accessed_time
+                    session_data['_accessed_time'] = now
+
+                    # Set the path if applicable
+                    if '_path' in session_data:
+                        self._path = session_data['_path']
+                    self.update(session_data)
+                    self.accessed_dict = session_data.copy()
+            finally:
+                self.namespace2.release_read_lock()
+            if timed_out:
+                self.invalidate()
```

Comment:
I guess this doesn't really do anything to cassandra, cassandra automatically expires the sessions correct ?